builtin/logical/database/rollback_test.go

```diff
@@ -0,0 +1,411 @@
+	data := map[string]interface{}{
+		"connection_url": connURL,
+		"plugin_name":    "postgresql-database-plugin",
+		"allowed_roles":  []string{"plugin-role-test"},
+		"username":       databaseUser,
+		"password":       defaultPassword,
+	}
+	resp, err := lb.HandleRequest(namespace.RootContext(nil), &logical.Request{
+		Operation: logical.UpdateOperation,
+		Path:      "config/plugin-test",
+		Storage:   config.StorageView,
+		Data:      data,
+	})
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("err:%s resp:%#v\n", err, resp)
+	}
+
+	// Create a role
+	data = map[string]interface{}{
+		"db_name":             "plugin-test",
+		"creation_statements": testRole,
+		"max_ttl":             "10m",
+	}
+	resp, err = lb.HandleRequest(context.Background(), &logical.Request{
+		Operation: logical.UpdateOperation,
+		Path:      "roles/plugin-role-test",
+		Storage:   config.StorageView,
+		Data:      data,
+	})
+	if err != nil || (resp != nil && resp.IsError()) {
+		t.Fatalf("err:%s resp:%#v\n", err, resp)
+	}
+
+	// Read credentials to verify this initially works
+	credReq := &logical.Request{
+		Operation: logical.ReadOperation,
+		Path:      "creds/plugin-role-test",
+		Storage:   config.StorageView,
+		Data:      make(map[string]interface{}),
+	}
+	credResp, err := lb.HandleRequest(context.Background(), credReq)
+	if err != nil || (credResp != nil && credResp.IsError()) {
+		t.Fatalf("err:%s resp:%v\n", err, credResp)
+	}
+
+	// Get a connection to the database plugin
+	pc, err := dbBackend.GetConnection(context.Background(),
+		config.StorageView, "plugin-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Alter the database password so it no longer matches what is in storage
+	_, _, err = pc.SetCredentials(context.Background(), dbplugin.Statements{},
+		dbplugin.StaticUserConfig{
+			Username: databaseUser,
+			Password: "newSecret",
+		})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Clear the plugin connection to verify we're no longer able to connect
+	err = dbBackend.ClearConnection("plugin-test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Reading credentials should no longer work
+	credResp, err = lb.HandleRequest(namespace.RootContext(nil), credReq)
+	if err == nil {
+		t.Fatalf("expected authentication to fail when reading credentials")
+	}
+
+	// Put a WAL entry that will be used for rolling back the database password
+	walEntry := &rotateRootCredentialsWAL{
+		ConnectionName: "plugin-test",
+		UserName:       databaseUser,
+		OldPassword:    defaultPassword,
+		NewPassword:    "newSecret",
+	}
+	_, err = framework.PutWAL(context.Background(), config.StorageView, rotateRootWALKey, walEntry)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertWALCount(t, config.StorageView, 1, rotateRootWALKey)
+
+	// Trigger an immediate RollbackOperation so that the WAL rollback
+	// function can use the WAL entry to roll back the database password
+	_, err = lb.HandleRequest(context.Background(), &logical.Request{
```

Comment:
I have a question - in real life, what would trigger the WalRollback function to be called? When I look at [the framework](https://github.com/hashicorp/vault/blob/master/sdk/framework/backend.go#L479), I see that it's called when it receives a rollback request (if I trace the code upwards.) What triggers this request? 

The reason I ask is because in the linked issue, the user restarted Vault and saw that the password was still out of sync. So, does that mean that the rollback method used here is called at startup? 

I was also thinking about this because I wondered if we should be considering using the `PeriodicFunc` on the back-end instead, which is called every 1 minute.

Reply:
Or I suppose, framed another way, would the user need to call the rollback endpoint?

Reply:
The `PeriodicFunc` and `WALRollbackFunc` for the backend are invoked on the same 1 minute interval by the [RollbackManager](https://github.com/hashicorp/vault/blob/master/vault/rollback.go#L19). See that the `PeriodicFunc` is invoked from the `RollbackOperation` request in [backend.go#L468](https://github.com/hashicorp/vault/blob/master/sdk/framework/backend.go#L468).

So, yes, the `WALRollbackFunc` will continue to work after a restart without any user intervention. The benefit to using `WALRollbackFunc` over `PeriodicFunc` here is that it's also supplied with WAL entries for the backend.

Hitting the unfortunate timing for this bug in the test was hard, so had to manually request for the immediate `RollbackOperation`.

Reply:
Ah gotcha! Cool cool!